### PR TITLE
CNX-9230 Revit Wall send

### PR DESCRIPTION
**Pull request by connorivy**
## Description & motivation

Adds the wall conversion. Specifically consolidated the existing logic into more digestible methods and added functionality to send curtain walls and stacked walls that didn't previously exist.

## Validation of changes:

Here is a working curtain wall sent with this conversion https://latest.speckle.dev/streams/85bc4f61c6/commits/49bc3fe387. You'll notice that you cannot select the entire wall, only the individual panels, but this was existing behavior.

## Checklist:

- [x] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [x] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [x] My commits are related to the pull request and do not amend unrelated code or documentation.
- [x] My code follows a similar style to existing code.
- [ ] I have added appropriate tests.
- [ ] I have updated or added relevant documentation.

### CODE REVIEW

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Extensions/ElementExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+    else
+    {
+      var typeFilter = new ElementIsElementTypeFilter(true);
+      var categoryFilter = new ElementMulticategoryFilter(
+        new List<BuiltInCategory>()
+        {
+          BuiltInCategory.OST_CLines,
+          BuiltInCategory.OST_SketchLines,
+          BuiltInCategory.OST_WeakDims
```
Comment (AlanRynne) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Extensions/ElementExtensions.cs:22]: I'm assuming these have been ported from the original logic. I'd be curious to know if there is any others we should include here.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/CurveArrArrayToSpecklePolycurveConverter.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    return polycurves;
+  }
+
+  private IEnumerable<DB.CurveArray> GetCurveArrays(DB.CurveArrArray curveArrArray)
+  {
+    for (var i = 0; i < curveArrArray.Size; i++)
+    {
+      yield return curveArrArray.get_Item(i);
+    }
+  }
```
Comment (AlanRynne) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/CurveArrArrayToSpecklePolycurveConverter.cs:33]: hahaha autodesk APIs... 🤮

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/RevitConverterToSpeckle.cs [hunk 3]
```diff
@@ -49,4 +49,19 @@ public Base Convert(object target)
...
+      return null;
+    }
+
+    if (_toSpeckle.ResolveInstance(objectType.Name) is IHostObjectToSpeckleConversion conversion)
+    {
+      return conversion;
+    }
+
+    return GetConversionForObject(objectType.BaseType);
+  }
```
Comment (AlanRynne) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/RevitConverterToSpeckle.cs:66]: ohhh, is this trying to resolve the converter including all parent type converters?

Neat! @BovineOx this is not a bad idea... could this be somehow moved into the default resolving logic or some new function of the converter (`ResolveInstance` and `ResolveInstanceWithInheritance`? name pending)

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs [hunk 4]
```diff
@@ -76,64 +89,73 @@ public override SOBR.RevitWall RawConvert(DB.Wall target)
...
     }
 
-    return curveArrays;
+    List<SOG.Polycurve> polycurves = _curveArrArrayConverter.RawConvert(profile);
+
+    if (polycurves.Count > 1)
+    {
+      // POC: we have been assuming that the first curve is the element and the rest of the curves are openings
+      // this isn't always true
+      speckleWall["voids"] = polycurves.Skip(1);
```
Comment (AlanRynne) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:158]: We know for a fact many times these are not even `voids`... 😅 we really should re-evaluate the whole wall conversions in revit to account for sketched walls properly

**Review by AlanRynne — APPROVED** [review]
I see no blockers but I left a comment for @BovineOx to look at 😉

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs (hunk 4 above)
Comment (connorivy) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:158]: I agree. We also need to stop using dynamic properties (in my opinion)

Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:158]: YES!! 🙌🏼 you've got my vote and @JR-Morgan's hahah

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/RevitConverterToSpeckle.cs (hunk 3 above)
Comment (BovineOx) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/RevitConverterToSpeckle.cs:66]: Yeah it needs some thought, maybe add a POC: comment to consider making this more accessible as a pattern - make a CNX and add it here

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs (hunk 4 above)
Comment (BovineOx) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/WallConversionToSpeckle.cs:158]: Can w add a ticket to review "re-evaluate the whole wall conversions in revit to account for sketched walls properly" and drop the CNX here plz?

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/HostedElementConversionToSpeckle.cs [hunk 5]
```diff
@@ -24,10 +25,12 @@ ISpeckleConverterToSpeckle converter
 
   public List<Base> GetHostedElementsConverted(Element host)
   {
-    var convertedHostedElements = new List<Base>();
-
-    var hostedElementIds = GetHostedElementIds(host);
+    return GetHostedElementsConvertedFromIds(host, host.GetHostedElementIds());
```
Comment (BovineOx) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/ToSpeckle/HostedElementConversionToSpeckle.cs:28]: Hosted elements were kind of OOS for Alpha so maybe this is fine for now but I think @clairekuang needs to be across this for other object relationships such as blocks, instancing etc...

Maybe add a POC: comment for now to the effect of, needs to be reviewed

**Review by BovineOx — CHANGES_REQUESTED** [review]
Mostly just adding comments and a couple of tickets

**Review by AlanRynne — APPROVED** [review]

**Pull request merged**